keras/layers/merge.py

```diff
@@ -27,19 +66,86 @@ def build(self, input_shape):
             raise ValueError('A merge layer should be called '
                              'on a list of at least 2 inputs. '
                              'Got ' + str(len(input_shape)) + ' inputs.')
-        if all([shape is None for shape in input_shape]):
-            return
-        # TODO: handle shapes with None entries.
-        input_shapes_set = set(input_shape)
-        if None in input_shapes_set:
-            input_shapes_set.remove(None)
-        if len(input_shapes_set) > 1:
-            raise ValueError('Only tensors of same shape can '
-                             'be merged by layer' + self.name +
-                             ' Got input shapes: %s' % input_shape)
+        batch_sizes = [s[0] for s in input_shape if s is not None]
+        batch_sizes = set(batch_sizes)
+        batch_sizes -= set([None])
+        if len(batch_sizes) > 1:
+            raise ValueError('Can not merge tensors with different '
+                             'batch sizes. Got tensors with shapes : ' +
+                             str(input_shape))
+        if input_shape[0] is None:
+            output_shape = None
+        else:
+            output_shape = input_shape[0][1:]
+        for i in range(1, len(input_shape)):
+            if input_shape[i] is None:
+                shape = None
+            else:
+                shape = input_shape[i][1:]
+            output_shape = self._compute_elemwise_op_output_shape(output_shape, shape)
+        if None not in input_shape and len(set(map(len, input_shape))) == 1:
+            self._permute_required = False
+        else:
+            self._permute_required = True
 
     def call(self, inputs):
-        return self._merge_function(inputs)
+        if self._permute_required:
+            reshaped_inputs = []
+            reshaped = False
+            for x in inputs:
+                x_ndim = K.ndim(x)
+                if x_ndim is None:
```

Comment:
This case is sufficiently irregular that we could simply raise a ValueError and required `ndim` to be defined. Seems like it would simplify the code quite a bit.